pallets/vrf/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! VRF Pallet
+//!
+//! Stores VRF output per block by the block author as well as the VRF inputs:
+//! `most_recent_relay_block_hash + most_recent_relay_slot_number`
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_application_crypto::ByteArray;
+use sp_consensus_babe::{digests::PreDigest, AuthorityId, Slot, Transcript, BABE_ENGINE_ID};
+use sp_consensus_vrf::schnorrkel;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+type MaybeRandomness = Option<schnorrkel::Randomness>;
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::Convert;
+
+	/// For the runtime to implement to expose cumulus data to this pallet
+	pub trait GetMostRecentVrfInputs<RelayHash, SlotNumber> {
+		fn get_most_recent_relay_block_hash() -> RelayHash;
+		fn get_most_recent_relay_slot_number() -> SlotNumber;
+	}
+
+	/// This trait tells us if the round changed in the current block
+	/// => new set of authorities to be put in storage
+	pub trait RoundChangedThisBlock {
+		fn round_changed_this_block() -> bool;
+	}
+
+	/// Exposes randomness in this pallet to the runtime
+	pub trait GetRandomness {
+		type Randomness;
+		fn get_last_randomness() -> Self::Randomness;
+		fn get_current_randomness() -> Self::Randomness;
+	}
+
+	/// Make VRF transcript
+	pub fn make_transcript<Hash: AsRef<[u8]>>(
+		relay_slot_number: Slot,
+		relay_block_hash: Hash,
+	) -> Transcript {
+		let mut transcript = Transcript::new(&BABE_ENGINE_ID);
+		transcript.append_u64(b"relay slot number", *relay_slot_number);
+		transcript.append_message(b"relay block hash", relay_block_hash.as_ref());
+		transcript
+	}
+
+	/// Pallet for storing and exposing VRF outputs by block authors
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The relay block hash type (probably H256)
+		type RelayBlockHash: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ Default
+			+ Copy
+			+ AsRef<[u8]>;
+		/// Gets the most recent relay block hash and relay slot number in `on_initialize`
+		type MostRecentVrfInputGetter: GetMostRecentVrfInputs<Self::RelayBlockHash, Slot>;
+		/// Convert account to VRF key, presumably via AuthorMapping instance
+		/// TODO: maybe TryConvert and map to error because fallible conversion
+		type AccountToVrfId: Convert<Self::AccountId, AuthorityId>;
```

Comment:
You will probably need to implement the Convert trait in our AccountId declaration (somwhere https://github.com/PureStake/moonbeam/blob/master/primitives/account/src/lib.rs probably). Is that fine for you? Else you can name this with a new trait name, and declare a struct that implements it in the runtime

Reply:
I would implement it for some struct declared in the runtime. It would use the `AuthorMapping` to get the mapping VrfId for the relevant AccountId (so we would have to add that backwards mapping for these keys).

I'm still thinking about how we should deal with no VRF key registered or to enforce the key must be registered for all SelectedCandidates (without adding that constraint to parachain-staking which may be the only way).

Reply:
Makes sense then! I though one could only implement Convert for a Self type